Sources/TuistCore/ValueGraph/ValueGraph.swift

```diff
@@ -0,0 +1,20 @@
+import Foundation
+import TSCBasic
+
+/// An directed acyclic graph (DAG) that Tuist uses to represent the dependency tree.
+struct ValueGraph: Equatable {
+    /// A dictionary where the keys are the paths to the directories where the projects are defined,
+    /// and the values are the projects defined in the directories.
+    let projects: [AbsolutePath: Project]
+
+    /// A dictionary where the keys are paths to the directories where the projects that contain packages are defined,
+    /// and the values are dictionaries where the key is the reference to the package, and the values are the packages.
+    let packages: [AbsolutePath: [String: Package]]
+
+    /// A dictionary where the keys are paths to the directories where the projects that contain targets are defined,
+    /// and the values are dictionaries where the key is the name of the target, and the values are the targets.
+    let targets: [AbsolutePath: [String: Target]]
```

Comment:
> Is the idea to incrementally update the types?

Good point. I didn't think about it but I think we can do it incrementally. Once we access the project targets through the graph, it should be straightforward to turn `Project` into a value type.

> I guess worth seeing how it turns out as the proposed plan has a few milestones which allow us to reflect on it and evaluate it before doing a final replacement.

Yeah, I think it's a combination of a graph, with the peculiarity that we have the representation of the targets extracted somewhere else so that we can do mutations without having to mutate the graph. Also, we have the one-to-many associations between projects and packages and targets.